### ISSUE TITLE
[Question] Why a Node.js add-on, instead of FFI?

### ISSUE THREAD

**Issue by modocache** [RLovelett/node-sourcekit#2]
This is awesome! Thanks for starting this, I'm looking forward to contributing.

I noticed that Python bindings exist in [swift/tools/SourceKit/bindings](https://github.com/apple/swift/tree/48bfec49fb4e97584caaa409a99fac48d47de6b0/tools/SourceKit/bindings). Honestly, I'm still trying to figure out how exactly they work. I _think_ they're simply Python wrappers for the C API to sourcekitd.

Thanks for [the great link on bridging C++ to Node.js](https://blog.scottfrees.com/getting-your-c-to-the-web-with-node-js). I noticed `node-sourcekit` is a Node.js add-on (option three from the article), whereas the existing Python bindings are FFI, or foreign function interfaces to the C code (option two from the article).

I'm curious to hear your thoughts on why to use an add-on instead of FFI. I was envisioning using FFI, for the following reasons:
1. FFI seems conceptually simpler, at least to me 😅 
2. I think it'd be great to push this code upstream into [swift/tools/SourceKit/bindings](https://github.com/apple/swift/tree/48bfec49fb4e97584caaa409a99fac48d47de6b0/tools/SourceKit/bindings). Because the Python bindings use FFI, and because I perceive FFI to be simpler, I think it'll be easier to merge JavaScript bindings into Swift trunk if they also used FFI.

I'm pretty new to SourceKit, JavaScript, and any sort of bridging, so just to be clear: this is a question, not developer snark. 😝  I'm genuinely curious what your thoughts are on FFI vs. add-on.

**Comment by aaroncrespo** [comment]
I'd love to try and get something like this upstreamed into swift. It might be a tricky maneuver
 considering we are adding additional dependencies. The whole ecosystem of editors with javascript backends helps the case hopefully. 

As for your other question I can't speak a ton I know when I initially researched this subject node-fifi's readme scared me off. Solutions like Swig.org or a Node.js plugin were more automatable(?) or maybe easier to maintain in terms of bridging to sourcekit. No need to map primitives or objects to JS -- if your plugin compiles your JS works and node does all the heavy lifting.

I may be wrong, but I think a plugin also might have less call overhead, in the grand scheme of things 
that might not count for much but in my head it makes sense. 💭 So sure why not. 

Like you JS is a means to an end to play with Editor tools there might be better ways of doing it above is just what I cobbled together over a few hours of research. @RLovelett did that in parallel and then actually did work so he can correct me if I got anything wrong.

**Comment by RLovelett** [comment]
> I noticed that Python bindings exist in swift/tools/SourceKit/bindings. Honestly, I'm still trying to figure out how exactly they work. I think they're simply Python wrappers for the C API to sourcekitd.

I think your right. This is just providing the C types into Python. Basically this is what this library is doing only its doing C API -> C++ v8 API -> Node.js ES6 wrapper.

**Caveat: scanning that really quickly and knowing virtually 0 Python.**

> I'm curious to hear your thoughts on why to use an add-on instead of FFI. I was envisioning using FFI, for the following reasons:
> 1. FFI seems conceptually simpler, at least to me 😅

At first glance I get your point.

A lot of "higher order" functions are just taken care of if the code stays in JavaScript (or Python). For instance, you get `map` and `flatMap` for monadic container types out of the box and you [do not have to implement them yourself](https://github.com/RLovelett/node-sourcekit/blob/be699f5972a65ae5aafce36a6351f9bb4be810ac/cpp/node-sourcekit.cpp#L27-L49). Also you do not have to worry about memory allocation and the like. All of those are immediate and obvious wins over C++ (IMO).

On the other hand you quickly realize FFI is literally interacting with the same interfaces that C++ is dealing with. [In other words, JavaScript ends up having to spend time casting and converting the native JavaScript types to the native types that SourceKit's C API accepts](http://blog.scottfrees.com/calling-native-c-dlls-from-a-node-js-web-app#calling-primelib-with-ffi). For example in C++ to send a string to `sourcekitd_request_dictionary_set_string` you end up having to cast the JavaScript string (e.g., `v8::String`) to `char*`. The casting in practice ends up looking something like `v8::Value` -> `v8::String` -> `Utf8String` -> `char*`. As best I can tell, without actually writing code to test my theory you end up having to do something similar with FFI. Only now you have no compiler type safety. Obviously that cast above is non-trivial and having a compiler's type-checker is valuable here (again at least to me).

> 1. I think it'd be great to push this code upstream into swift/tools/SourceKit/bindings. Because the Python bindings use FFI, and because I perceive FFI to be simpler, I think it'll be easier to merge JavaScript bindings into Swift trunk if they also used FFI.

I would love that too. If you think that FFI is the way to go to achieve that, then to say the very least I can on the subject, I am interested in migrating approaches.

That having been said, given how much of Swift is actually written in C++ I cannot imagine C++ code is going to scare off the project maintainers of Swift. That of course is purely speculation on my part though.

## Bottom Line

I'm open to switching to FFI.

After playing around with the code and seeing how _relatively_ easy the v8 engine interfaces with the SourceKit C API I'd at least like to see a comparison of the FFI code with the C++ add-on code. I think if its a wash and both are equally opaque, as I imagine they will be, then the deciding on a path forward just becomes a matter of language preference and ability to eventually be absorbed upstream.

Any one interested in choosing a _simple_ SourceKit C API and wrapping it using FFI?

**Comment by modocache** [comment]
> node-ffi's readme

Oh wow -- until you mentioned this I had assumed FFI was built into some sort of standard library for Node.js, like Python's `ctypes` module. I looks like that [isn't the case](https://github.com/node-ffi/node-ffi). That definitely changes my viewpoint. On the contrary, I guess [nan](https://github.com/nodejs/nan) is an "official" part of Node.js.

> I may be wrong, but I think a plugin also might have less call overhead, in the grand scheme of things  that might not count for much but in my head it makes sense.

I imagine the indexing and parsing done by SourceKit itself will be the performance bottleneck for any reasonably sized Swift package. The call overhead seems like it would be negligible, but I might be wrong.

> I'm open to switching to FFI.

On the contrary, you and @aaroncrespo have convinced me an add-on via nan is the way to go! 😝  Thanks a ton for all the feedback.
